launchdarkly/ios-eventsource
# Safer Event Dispatch

LDEventSource sometimes passes an event with no event or data to the SDK. 
There are 2 causes:
1. The LDEventSource calls `dispatch_async` on a new event. Sometimes the dispatch happens immediately, and other times after a delay. When it’s delayed, the event gets replaced with a new empty event.
2. iOS can break a single event into two calls to the NSURLSession delegate method `didReceiveData`. This was never planned for in the original code.

This change corrects both issues by refactoring LDEventSource to use a data accumulator and event parser instead of doing these itself. The data accumulator accumulates data until it is ready to be parsed into an event. An empty line from the `clientstream` signals the end of an event, and the data accumulator uses this signal to measure when the data it has received has an empty line as the last line of the data.
Once the data accumulator senses that the data is ready to be parsed, LDEventSource calls the LDEventParser with the accumulated data. The parser creates the event and LDEventSource dispatches it. If there is more data available, the parser tells the LDEventSource that it has more data to parse. The LDEventSource then recursively calls its `parseEventString` method on the remaining data, which repeats the process.

## Test
### Test setup
1. Use the sample app provided.
2. Configure a web proxy to monitor communication between the SDK and LD servers
3. Login to the LD staging Dashboard. Navigate to the appropriate Project and Environment. A new flag `really-long-string` has been added. By turning `really-long-string` `ON` you can more reliably cause iOS to split a single SSE Event into multiple calls to `didReceiveData`.
4. In the sample app set the following breakpoints configured to log a message to the console and automatically continue:
4a. LDEventSource URLSession: dataTask: didReceiveData: on the test `isReadyToParseEvent` (line 158). Message to log: 
>LDEventSource didReceiveData isReadyToParseEvent:@[self.dataAccumulator isReadyToParseEvent]@

Setting this will cause a message in the console each time iOS calls `didReceiveData`. When `NO` the data will be accumulated, but not parsed. When `YES` the data will be accumulated and then parsed into an event.
4b. ViewController featureFlagDidUpdate: on the `if([key isEqualToString:])` (line 114). Message to log:
>hello-ios featureFlagDidUpdate:@key@

4c. ViewController userDidUpdate: on `[self checkFeatureValue]`. Message to log: 
>hello-ios userDidUpdate

4d.ViewController userUnchanged: on `[self checkFeatureValue]`. Message to log: 
>hello-ios userUnchanged

The breakpoints set at 4b, 4c, & 4d help validate that an event was created that caused the SDK to call a client delegate method.

### Test steps
1. On the Dashboard, turn off the following flags: `hello-ios-boolean`, `hello-ios-number`, `hello-ios-double`, `hello-ios-string`, `really-long-string`. Turn on the following flags: `hello-ios-array`, `hello-ios-dictionary`.
2. Launch the sample app. Verify the SDK connects to the `clientstream`. Verify the sample app shows the flag values set in step 1 above.
3. In the Xcode console, verify the LDEventSource didReceiveData at least once with `isReadyToParseEvent` `YES`. Verify the SDK called `hello-ios` `userDidUpdate` or `userUnchanged` once.
4. Kill the sample app.
### Verify `put` event
5. On the dashboard, turn on `hello-ios-boolean`.
6. Launch the sample app. Verify the sample app shows `hello-ios-boolean` `YES`.
7. In the Xcode console, Verify LDEventSource `didReceiveData` was called only once ready to parse event. Verify `hello-ios` `featureFlagDidUpdate` was called with `hello-ios-boolean`. Verify `hello-ios` `userDidUpdate` was called.
8. Kill the sample app.
9. On the Dashboard, turn on `really-long-string` and turn off `hello-ios-boolean`.
10. Launch the sample app. Verify the sample app shows `hello-ios-boolean` `NO`.
11. In the Xcode console, Verify LDEventSource `didReceiveData` was called at least twice, only once ready to parse event. Verify `hello-ios` `featureFlagDidUpdate` was called with `hello-ios-boolean` and `really-long-string`. Verify `hello-ios` `userDidUpdate` was called.
### Verify `patch` event
12. On the Dashboard, turn off `really-long-string`.
13. In the Xcode console, verify LDEventSource `didReceiveData` was called only once ready to parse event. Verify `hello-ios` `featureFlagDidUpdate` was called with `really-long-string`. Verify `hello-ios` `userDidUpdate` was called.
14. On the Dashboard, turn on `really-long-string`.
15. In the Xcode console, verify `LDEventSource` `didReceiveData` was called at least twice, only once ready to parse event. Verify `hello-ios` `featureFlagDidUpdate` was called with `really-long-string`. Verify `hello-ios` `userDidUpdate` was called.
